File: Lab/Project_1/Prob1/py_version/1_1.py
from collections import deque

N = 100010
M = 2 * N
graph = [-1] * N
e = [0] * M
ne = [0] * M
distance = [-1] * N
idx = 0


def add(a,b):
    global idx
    idx += 1
    e[idx] = b
    ne[idx] = graph[a]
    graph[a] = idx


def bfs():
    queue = deque()
    queue.append(1) # 队列中存储着节点编号
    distance[1] = 0

    while queue:
        node = queue.popleft()
        d = distance[node]
        if node == n:
            return d
        cur = graph[node]
        while cur != -1:
            j = e[cur]
            if distance[j] == -1:
                queue.append(j)
                distance[j] = d + 1
            cur = ne[cur]
    return -1



if __name__ == "__main__":
    n,m = map(int,input().split())
    while m > 0:
        a,b = map(int,input().split())
        add(a,b)
        m -= 1

    dis = bfs()
    print(dis)

File: Lab/Project_1/Prob1/cpp_version/1_1.cpp
#include <iostream>
#include <cstring>

using namespace std;
const int N = 100010;
int q[N], hh = 0, tt = -1;
int g[N], dist[N], e[N], ne[N], idx;
int n, m;

void add(int a, int b)
{
    idx ++, e[idx] = b, ne[idx] = g[a], g[a] = idx;
}

int bfs()
{
    q[++tt] = 1;
    dist[1] = 0;

    while (hh <= tt)
    {
        int node = q[hh++];
        if (node == n)  return dist[n];

        for (int cur = g[node]; cur != -1; cur = ne[cur])
        {
            int j = e[cur];
            if (dist[j] == -1)  q[++tt] = j, dist[j] = dist[node] + 1;
        }
    }

    return -1;
}


int main()
{
    memset(g, -1, sizeof g);
    memset(dist, -1, sizeof dist);
    cin >> n >> m;
    int a, b;
    while (m --)
    {
        cin >> a >> b;
        add(a, b);
    }

    printf("%d\n", bfs());

    return 0;
}